security/acme-client/src/opnsense/mvc/app/library/OPNsense/AcmeClient/LeAccount.php

```diff
@@ -0,0 +1,263 @@
+     */
+    public function generateKey()
+    {
+        // Collect account information
+        $account_conf_dir = self::ACME_BASE_ACCOUNT_DIR . '/' . (string)$this->config->id . '_' . $this->environment;
+        $account_conf_file = $account_conf_dir . '/account.conf';
+        $account_key_file = $account_conf_dir . '/account.key';
+        $account_json_file = $account_conf_dir . '/account.json';
+        $account_ca_file = $account_conf_dir . '/ca.conf';
+        $acme_conf = array();
+        $acme_conf[] = "CERT_HOME='" . self::ACME_HOME_DIR . "'";
+        $acme_conf[] = "LOG_FILE='" . self::ACME_LOG_FILE . "'";
+        $acme_conf[] = "ACCOUNT_KEY_PATH='" . $account_key_file . "'";
+        $acme_conf[] = "ACCOUNT_JSON_PATH='" . $account_json_file . "'";
+        $acme_conf[] = "CA_CONF='" . $account_ca_file . "'";
+        if (!empty((string)$this->config->email)) {
+            $acme_conf[] = "ACCOUNT_EMAIL='" . (string)$this->config->email . "'";
+        }
+
+        // Store some values for later re-use
+        $this->account_conf_file = $account_conf_file;
+
+        // Create account configuration file
+        if (!is_dir($account_conf_dir)) {
+            mkdir($account_conf_dir, 0700, true);
+        }
+        file_put_contents($account_conf_file, (string)implode("\n", $acme_conf) . "\n");
+        chmod($account_conf_file, 0600);
+
+        // Check if account key already exists both in filesystem and in config
+        if (!is_file($account_key_file) || empty((string)$this->config->key)) {
+            LeUtils::log_debug('creating account key for ' . (string)$this->config->name, $this->debug);
+
+            // Check if we have an account key in our configuration
+            if (!empty((string)$this->config->key)) {
+                LeUtils::log_debug('exporting existing account key to filesystem for ' . (string)$this->config->name, $this->debug);
+                // Write key to disk
+                file_put_contents($account_key_file, (string)base64_decode((string)$this->config->key));
+                chmod($account_key_file, 0600);
+                return true;
+            } else {
+                LeUtils::log_debug('generating a new account key for ' . (string)$this->config->name, $this->debug);
+                // Preparation to run acme client
+                $proc_env = $this->acme_env; // env variables for proc_open()
+                $proc_env['PATH'] = $this::ACME_ENV_PATH;
+                $proc_desc = array(  // descriptor array for proc_open()
+                    0 => array("pipe", "r"), // stdin
+                    1 => array("pipe", "w"), // stdout
+                    2 => array("pipe", "w")  // stderr
+                );
+                $proc_pipes = array();
+
+                // Run acme client to generate a account key
+                $acmecmd = '/usr/local/sbin/acme.sh '
+                  . '--createAccountKey '
+                  . implode(' ', $this->acme_args) . ' '
```

Comment:
in many places this looks prone to shell command injection especially since arguments are given in pairs and can't be easily escaped afterwards

Reply:
@fichtner Thanks, is there an easy way to fix this? Maybe a built-in PHP function to sanitize the arguments?

Reply:
@fichtner Do you think https://github.com/opnsense/plugins/pull/2028/commits/4036e452f46a189d139e1b21b05b6be4a2298d57 would improve the situation?

Reply:
sorry for the delay. no, this is too late, you need to escape all arguments of the array cleanly in a way the shell would separate them... as far as I can see you add "-commandswitch argument" to the array but these are two separate entities for escaping.

the following does the job prior to invoke but you need to split the pushing of acme_args correctly

https://github.com/opnsense/core/blob/master/src/etc/inc/util.inc#L1123

Reply:
See e.g. https://github.com/opnsense/core/blob/86872d4170a442132ef2f05d394e9a6d21306376/src/www/diag_testport.php#L73